.github/workflows/buildAndTest.yml

```diff
@@ -119,11 +107,14 @@ jobs:
           fetch-depth: 2
           submodules: "true"
 
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.11'
```

Comment:
Not 3.11 but I would like at least 3.9 for [Type Hinting Generics in Standard Collections](https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections) and ideally 3.10 for [writing union types as `X | Y`](https://peps.python.org/pep-0604/). To be honest, I do not see the value of being behind current main release of python (soon to be 3.12). We're talking about MLIR here, which is already bleeding edge technology and has basically no notion of semver.

Reply:
I agree for modern Python and C++23 at the same time. ;-)

Reply:
I'm personally always running the latest version, but in a public project we need to be a bit more reasonable. As a rule of thumb, I'd stay at most at whatever version ships with the oldest fully supported Ubuntu LTS (3.8.2 at the moment). The more complicated you make it for users to run your software, the fewer of them will stick with it. And if you expect it to be deployed within any company with any sort of software policy and IT department, you can forget about it.

Reply:
But is it our goal to promote implicitly technical debt by supporting almost 6 year old LTS?
Just advise to use a VM to run in an antique shop or to move to a company interested by software? :-)

Reply:
> but in a public project we need to be a bit more reasonable.

This public project depends on MLIR which has no semver (as I mentioned). As I also mentioned, I being basically the core maintainer of upstream Python bindings, am always pushing to bump the supported Python version because Python has no breaking changes between 3.8 and 3.12 (so upgrading is "free"). So in summary: IMHO, it doesn't make sense to be conservative downstream of an unstable upstream.

Reply:
> But is it our goal to promote implicitly technical debt by supporting almost 6 year old LTS? Just advise to use a VM to run in an antique shop or to move to a company interested by software? :-)

You mean like ours? Because we do run 20.04 LTS in all our workstations 😂 

> This public project depends on MLIR which has no semver (as I mentioned). As I also mentioned, I being basically the core maintainer of upstream Python bindings, [am always pushing to bump the supported Python version](https://github.com/llvm/llvm-project/pull/68009) because Python has no breaking changes between 3.8 and 3.12 (so upgrading is "free"). So I can reduce the version here but in a matter of months upstream (me) will bump the required version anyway.
> 
> So in summary: IMHO, it doesn't make sense to be conservative downstream of an unstable upstream.

I may be mistaken here, but I thought LLVM required Python >=3.6, I'd expect everything to compile and run with 3.6 (tangentially, I bet that was 18.4 LTS included version, which went out of full support earlier this year). Ultimately, I'm just making a suggestion here. It's good practice to be (reasonably) conservative with dependency versions when your software is intended for wider adoption. And just to be absolutely clear, that's not my preference (I update the dependencies of all my personal projects every time I do anything with them, and I run Debian Sid in my PC), that's just a fact of life 🤷

Reply:
I don't know about LLVM but MLIR currently supports >=3.8 (but not for long).

Reply:
Being part of LLVM, I expected requirements to be aligned. TIL.

Reply:
> You mean like ours? Because we do run 20.04 LTS in all our workstations 😂

Oh, I thought you were talking about external users. :-)
Just ask IT to upgrade to 22.04 and/or to install whatever Python version?